### PR TITLE
Two fixes for Segment's mapping of "traits" to FullStory user variable names.

### PR DESCRIPTION
**Pull request by fka3**
1) When users pass compound objects (arrays or non-Date objects), don't mess with the
   key name.  (Existing behavior makes the key become "undefined"!)  FullStory isn't
   going to *accept* the compound, but this way it can whine about the correctd name.
2) If a trait name is already type-tagged, don't mess with that.  In particular, don't
   double a correct type (foo_int shouldn't become foo_int_int) and don't sweat a
   mismatch (foo_int shouldn't become foo_int_str if given a string value).  Again,
   FullStory may whine if there's a mismatch, but the whine is correct and the name
   in FullStory's error message should be what the user expected.

### CODE REVIEW

**Comment by fka3** [comment]
This change was inspired by a user passing a trait like { subscription: { status: "premier", first_seen: "2016-02-01" } } into analytics.identify.  Because the value of subscription isn't one of the supported-by-FullStory types, the key conversion failed badly and we at FullStory got { undefined: { ... } }, leading to an error message about "undefined" being a bad variable name (which it is, but the user had no idea what we were talking about!) 

(There are other problems with the type inference, too, for example that { lifetime_value: 345.00 } would become lifetimeValue_int, even if it might after another purchase no longer be a round number of dollars, creating a different lifetimeValue_real, and maybe going back to _int again later.  That's not addressed here.)

File: lib/index.js [hunk 1]
```diff
@@ -75,13 +77,37 @@ FullStory.prototype.identify = function(identify) {
...
+    }
+  }
+
   key = camel(key);
-  if (is.string(value)) return key + '_str';
-  if (isInt(value)) return key + '_int';
-  if (isFloat(value)) return key + '_real';
-  if (is.date(value)) return key + '_date';
-  if (is.boolean(value)) return key + '_bool';
+  if (is.string(value)) {
```
Comment (f2prateek) [inline lib/index.js]: Please don't reformat our code, or send it as a separate PR if needed. Just adds noise to this one.

Comment (fka3) [inline lib/index.js]: It's not a bare reformat.  Instead of returning a type-qualified key (or not, in the case that none of the if's match, as was the problem resulting in our user getting a variable literally named "undefined"), the function is now both computing a new key AND setting it (line 110).

I broke it out this way because at first I wanted to flatten the object; it might be that instead of being 1-to-1, a single trait key/value would set many entries in results (e.g. key.subkey1 = value.subkey1, key.subkey2 = value.subkey2, etc.)  We backed away from that, however, because it isn't clear that flattening is good, either (in particular, it loses the collectiveness of an address, for example).

So if you'd rather, I can restore the 1-to-1 semantic while we ponder what object values should mean for our indexing and searching... I'm not entirely sure which formulation is "better," so I can be swayed by an argument either way, but it's not a casual reformat.

**Comment by fka3** [comment]
@f2prateek, reaction with the smaller delta?

File: test/index.test.js [hunk 2]
```diff
@@ -110,6 +110,25 @@ describe('FullStory', function() {
...
+
+      it('should map user objects properly', function() {
+        analytics.identify('id3', { account: { level: 'premier', avg_annual: 30000 } });
+        analytics.called(window.FS.identify, 'id3', { account: { level: 'premier', avg_annual: 30000 } });
+      });
+
+      it('should respect existing type tags', function() {
+        // All of these would break FS if passed; they prove that if the user has a
+        // variable of type X, but pass a type Y, we use their name for the variable.
+        analytics.identify('id3', { my_int_str: 17, my_str_int: 'foo', my_int_date: 3,
```
Comment (f2prateek) [inline test/index.test.js]: why are people calling identify with such weirdly named properties? `my_date` should suffice without needing the type in the key name. Can you share a more realistic example?

Comment (fka3) [inline test/index.test.js]: What I want is that if users DO provide type tagging, FullStory style (see http://help.fullstory.com/develop-js/setuservars), that Segment shouldn't mess that up by, for example, doubling the tag.  Or if they send a tag but have a mismatch, I want that tag and the mismatch to pass through, and not be "fixed" here.

If you want a "realistic" example, consider something like a total_spend variable, which is tracking the money spent by a user over their lifetime on the site.  Most of the time, that's going to be a real, like 34.56.  And if the user sends you guys a trait like { total_spend: 34.56 }, this code is going to send FullStory { total_spend_real: 34.56 }, which is great.

Every now and then, however, the user's spend is just by coincidence going to be a round number of dollars.  And because JavaScript doesn't differentiate Numbers, this code is going to send FullStory { total_spend_int: 112 } instead, creating a SECOND user variable, with a different type, in our system.  The goal here would be to allow such a user to send you total_spend_real and have it, by specifying a type (optionally), work regardless of that exact value.

Comment (f2prateek) [inline test/index.test.js]: But what happens when a user sends `total` (a specced value by us and used by a lot of integration) which might be a  any numeric type (float, int) - it looks like full story will have issues with this and require our users to send `total_real` which defeats of us defining a spec.

The better way is to recommend customers to force a type when they need it - instead of specifying it in the key. In either case they have change their code, but the forcing the type is more portable between integrations than defining a key with the type.

Comment (fka3) [inline test/index.test.js]: I'd argue the better way would be to take your fixed variables like total and send them to FullStory as total_real (on the grounds that a real can hold an int, but not the reverse).  You already special case displayName and email for example (where we don't need type information because they're special to us also); it'd be the same pattern.

But sure, for argument's sake, let's go with your suggestion to encourage the customer to force a type.  In this case, let's go to average_monthly_spend if total has meaning for you, and by sheer coincidence the average monthly spend is exactly $35.00.  How do they force Javascript pass that as a real not an int, when Javascript only supports "number?"

Comment (fka3) [inline test/index.test.js]: Also, at FullStory, we imagine we might expand our set of types.  We might want "URL" as a kind of string, for example, and we'd index things like the protocol, host, path, and query bits separately so you could do partial-match queries.  Or we might offer _money, which would be a kind of real that we understood had currency semantics, and we might do some special analytics about that because people think money is especially important.

Assuming we do have such semantically aware types, we'd like a way for users to be able to get that information through segment to use the features!

File: lib/index.js [hunk 3]
```diff
@@ -76,12 +76,24 @@ FullStory.prototype.identify = function(identify) {
...
+  }
+
+  // No tag found, try to infer one from the value.
   key = camel(key);
   if (is.string(value)) return key + '_str';
   if (isInt(value)) return key + '_int';
   if (isFloat(value)) return key + '_real';
   if (is.date(value)) return key + '_date';
   if (is.boolean(value)) return key + '_bool';
+  return key;  // final fall-through, bad FullStory type, but don't mess with the key
```
Comment (fka3) [inline lib/index.js]: Honestly, this (and not the already-tagged keys we've been discussing) is the change I care most about, because the result of users passing an object variable is that we get a dictionary like { undefined: { ... } } passed to us.

So if you want me to split this and the "should map ... properly" tests out into one PR, and we can discuss the other change separately, let me know.

**Comment by fka3** [comment]
Absent any further discussion here, I'm going to go ahead and break out the "undefined value" portion and send that as a separate PR.  Hopefully that's not controversial.  We can regroup about the double or incorrect tag parts later (where I still think it's a helpful change, but we can have that discussion separately.)
